### PR TITLE
[ConstraintSystem] Fix the ordering of functions with optional parame…

### PR DESCRIPTION
**Pull request by rudkx**
…ters.

Treat non-optional generic parameters as being more specialized than
optional generic parameters, and penalize any solutions that involve
generic arguments that are themselves Optional.

By doing these things, we can remove the special-cased code for the
two overloads of `??` in the stdlib, instead treating the `(T?, T)`
overload as better than the `(T?, T?)` overload except where a user
actually passes an optionally-typed value as the second parameter.

Fixes: rdar://problem/19748710

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please smoke test

File: test/Constraints/diag_ambiguities.swift [hunk 1]
```diff
@@ -29,11 +29,12 @@ C(g) // expected-error{{ambiguous use of 'g'}}
...
-func rdar29691909_callee(_ o: AnyObject?) -> Any? { return o } // expected-note {{found this candidate}}
-func rdar29691909_callee(_ o: AnyObject) -> Any { return o } // expected-note {{found this candidate}}
+func rdar29691909_callee(_ o: AnyObject?) -> Int? { fatalError() }
+func rdar29691909_callee(_ o: AnyObject) -> Int { fatalError() }
 
-func rdar29691909(o: AnyObject) -> Any? {
-  return rdar29691909_callee(o) // expected-error{{ambiguous use of 'rdar29691909_callee'}}
+func rdar29691909(o: AnyObject) -> Int {
+  let r = rdar29691909_callee(o)
+  return r
```
Comment (jrose-apple) [inline test/Constraints/diag_ambiguities.swift]: This is a different test. Please leave the original one intact, where there's a same-type constraint.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
As my latest test update shows, the change for `Any` vs. `Any?` introduces another change, which is that we now prefer `f<T>(x: T)` over `f(Any)`, which seems to me like a reasonable choice since you can select the second overload by doing `f(arg as Any)` but under the old scheme you can never select the `f<T>(x: T)` variant.

@DougGregor what do you think?

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Pull request merged**